editor/editor_dock_manager.cpp

```diff
@@ -52,6 +53,86 @@ enum class TabStyle {
 
 EditorDockManager *EditorDockManager::singleton = nullptr;
 
+bool EditorDockDragHint::can_drop_data(const Point2 &p_point, const Variant &p_data) const {
+	return can_drop_dock;
+}
+
+void EditorDockDragHint::drop_data(const Point2 &p_point, const Variant &p_data) {
+	if (can_drop_dock) {
+		Control *dragged_dock = dock_manager->_get_dock_tab_dragged();
+		int drop_tab = dock_manager->dock_slot[occupied_slot]->get_tab_bar()->get_closest_tab_idx_to_point(p_point);
+		dock_manager->_move_dock(dragged_dock, dock_manager->dock_slot[occupied_slot], drop_tab);
+	}
+}
+
+void EditorDockDragHint::set_slot(EditorDockManager::DockSlot p_slot) {
+	occupied_slot = p_slot;
+}
+
+void EditorDockDragHint::_notification(int p_what) {
+	switch (p_what) {
+		case EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED: {
+			if (EditorSettings::get_singleton()->check_changed_settings_in_group("interface/theme")) {
+				dock_drop_highlight->set_corner_radius_all(EDSCALE * EDITOR_GET("interface/theme/corner_radius").operator int());
+				if (is_visible_in_tree()) {
+					queue_redraw();
+				}
+			}
+		} break;
+
+		case NOTIFICATION_THEME_CHANGED: {
+			valid_drop_color = get_theme_color(SNAME("accent_color"), EditorStringName(Editor));
+			invalid_drop_color = get_theme_color(SNAME("error_color"), EditorStringName(Editor));
+		} break;
+
+		case NOTIFICATION_MOUSE_ENTER:
+		case NOTIFICATION_MOUSE_EXIT: {
+			bool new_mouse_inside = p_what == NOTIFICATION_MOUSE_ENTER;
+			if (new_mouse_inside != mouse_inside) {
+				mouse_inside = new_mouse_inside;
+				queue_redraw();
+			}
+		} break;
+
+		case NOTIFICATION_DRAG_BEGIN: {
+			Control *dragged_dock = dock_manager->_get_dock_tab_dragged();
+			if (!dragged_dock) {
+				return;
+			}
+
+			// TODO: Update logic when GH-106503 is merged to use flags.
+			can_drop_dock = is_layout_horizontal ? bool(dragged_dock->call("_can_dock_horizontal")) : true;
+			dock_drop_highlight->set_border_color(can_drop_dock ? valid_drop_color : invalid_drop_color);
+		} break;
+		case NOTIFICATION_DRAG_END: {
+			dock_manager->_dock_drag_stopped();
+			can_drop_dock = false;
+			hide();
+		} break;
+
+		case NOTIFICATION_DRAW: {
+			// Draw highlights around docks that can be dropped.
+			if (mouse_inside) {
+				Rect2 dock_rect = Rect2(Point2(), get_size()).grow(2 * EDSCALE);
+				draw_style_box(dock_drop_highlight, dock_rect);
+			}
+		} break;
+	}
+}
+
+EditorDockDragHint::EditorDockDragHint() {
+	dock_manager = EditorDockManager::get_singleton();
+
+	set_as_top_level(true);
+	dock_drop_highlight.instantiate();
+	dock_drop_highlight->set_draw_center(false);
```

Comment:
Maybe this should draw a partially transparent center so it is more clear that it isn't for the dock's content?